### PR TITLE
Changing graph identifiers from content keys to numbers

### PR DESCRIPTION
**Pull request by joeyslater**
# ↪️ Pull Request

Changing the content key aware identifiers (strings) to auto incrementing numbers. This will unlock further changes as replacing the adjacency lists. 

* created an opaque node id type
* instead of passing nodes around, passing node ids
* updated where certain add actions took place 

Results:

**Before**
| Graph Type   | Nodes |  Edges |  Serialize w/ Write | Deserialize w/ Read | Serialized Memory | Total Cold Build
| ------------- | ------------- |  ------------- |  ------------- | ------------- | ------------- |  ------------- |
| BundleGraph  | 48348  | 127683 | 1861 ms | 2509 ms | 68.96 mb | 741.96s
| RequestGraph  | 92336  | 715971 | 11133 ms | 5102 ms | 188.32 mb

**With Numeric Ids**
| Graph Type   | Nodes |  Edges |  Serialize w/ Write | Deserialize w/ Read | Serialized Memory | Total Cold Build
| ------------- | ------------- |  ------------- |  ------------- | ------------- | ------------- |  ------------- |
| BundleGraph  | 48348  | 127683 | 1361 ms (-26%)  | 1758 ms (-30%) | 59.24 mb (-14%) | 588.86s (-20%)
| RequestGraph  | 92344  | 716001 | 7713 ms (-31%)  | 2996 ms (-41%) | 102.43 mb (-45%)

## 🚨 Test instructions
* Running the current suite of unit and integration tests
* [x] add unit tests for the content graph
 
## ✔️ PR Todo

- [x ] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.52s | +27.00ms |
| Cached | 439.00ms | -28.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 553.00ms | -33.00ms 🚀 |
| dist/legacy/index.76ce4b0c.js | 1.12kb | +0.00b | 722.00ms | -478.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 717.00ms | +58.00ms ⚠️ |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 18.51s | +1.09s ⚠️ |
| Cached | 665.00ms | +56.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 477.81kb | +0.00b | 7.97s | +620.00ms ⚠️ |
| dist/PermalinkedComment.0a19c7e4.js | 4.24kb | +0.00b | 865.00ms | -87.00ms 🚀 |
| dist/UserProfile.b2960402.js | 1.78kb | +0.00b | 871.00ms | -81.00ms 🚀 |
| dist/NotFound.f3269ae7.js | 542.00b | +0.00b | 291.00ms | +43.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 97.00ms | -8.00ms 🚀 |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.00m | -23.36s 🚀 |
| Cached | 2.94s | +100.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.e48a72e5.js | 2.46mb | +6.00b ⚠️ | 1.04m | -10.67s 🚀 |
| dist/pdfRenderer.f5a3f4a7.js | 1.11mb | +0.00b | 38.67s | -34.38s 🚀 |
| dist/editorView.ea13a2b7.js | 588.08kb | +0.00b | 51.87s | -34.00s 🚀 |
| dist/EmojiPickerComponent.c4358f76.js | 139.84kb | +0.00b | 11.21s | +1.07s ⚠️ |
| dist/Toolbar.0c83bb49.js | 100.68kb | +0.00b | 15.48s | -1.40s 🚀 |
| dist/media-viewer.ebcb5e1e.js | 73.43kb | +0.00b | 11.28s | -2.25s 🚀 |
| dist/card.df70c9b7.js | 54.67kb | +0.00b | 10.89s | +1.32s ⚠️ |
| dist/card.1d931b4c.js | 52.43kb | +0.00b | 17.23s | +3.29s ⚠️ |
| dist/component.ad27194c.js | 31.05kb | +0.00b | 1.89s | -394.00ms 🚀 |
| dist/dropzone.3ae9a030.js | 16.15kb | +0.00b | 11.83s | -3.11s 🚀 |
| dist/ui.15dd7acd.js | 14.30kb | +0.00b | 10.64s | +1.42s ⚠️ |
| dist/component.677fad45.js | 6.23kb | +0.00b | 3.73s | -337.00ms 🚀 |
| dist/card.3a4028bd.js | 5.74kb | +0.00b | 11.27s | -857.00ms 🚀 |
| dist/media-viewer.f15e56f4.js | 4.01kb | +0.00b | 11.37s | +971.00ms ⚠️ |
| dist/EmojiPickerComponent.4991ddd4.js | 3.68kb | +0.00b | 9.25s | +2.04s ⚠️ |
| dist/dropzone.d1e8004d.js | 3.46kb | +0.00b | 11.79s | +875.00ms ⚠️ |
| dist/index.ece3a13e.css | 3.46kb | +0.00b | 19.82s | +1.49s ⚠️ |
| dist/Modal.08282ab6.js | 3.23kb | +0.00b | 3.42s | -473.00ms 🚀 |
| dist/clipboard.6a6e5fbd.js | 3.00kb | +0.00b | 11.83s | -3.12s 🚀 |
| dist/16.b6ca7fd0.js | 2.50kb | +0.00b | 1.89s | -394.00ms 🚀 |
| dist/ResourcedEmojiComponent.5046c0ff.js | 2.19kb | +0.00b | 10.64s | +1.42s ⚠️ |
| dist/card.2888338b.js | 2.15kb | +0.00b | 11.83s | +1.42s ⚠️ |
| dist/date.a69000aa.js | 1.97kb | +0.00b | 4.88s | -287.00ms 🚀 |
| dist/16.a9ea9b41.js | 1.86kb | +0.00b | 5.35s | +482.00ms ⚠️ |
| dist/browser.8883c452.js | 1.85kb | +0.00b | 13.56s | -3.20s 🚀 |
| dist/16.89d45d15.js | 1.80kb | +0.00b | 3.94s | -496.00ms 🚀 |
| dist/workerHasher.d607e0b8.js | 1.77kb | +0.00b | 17.36s | +3.26s ⚠️ |
| dist/list-number.d4568b49.js | 1.69kb | +0.00b | 5.02s | -698.00ms 🚀 |
| dist/link.5be155be.js | 1.54kb | +0.00b | 5.01s | -702.00ms 🚀 |
| dist/16.07ccad90.js | 1.51kb | +0.00b | 4.59s | -498.00ms 🚀 |
| dist/16.00f25203.js | 1.47kb | +0.00b | 3.94s | -496.00ms 🚀 |
| dist/16.02056ae6.js | 1.47kb | +0.00b | 4.59s | -497.00ms 🚀 |
| dist/16.b7da7bde.js | 1.45kb | +0.00b | 1.89s | -394.00ms 🚀 |
| dist/16.00100306.js | 1.44kb | +0.00b | 4.76s | -259.00ms 🚀 |
| dist/16.bdd61d2d.js | 1.41kb | +0.00b | 4.75s | -265.00ms 🚀 |
| dist/16.fa5f411c.js | 1.36kb | +0.00b | 5.35s | +482.00ms ⚠️ |
| dist/16.2a21d3f4.js | 1.34kb | +0.00b | 1.89s | -394.00ms 🚀 |
| dist/16.c9db7e6b.js | 1.32kb | +0.00b | 4.76s | -255.00ms 🚀 |
| dist/layout.81f2196c.js | 1.28kb | +0.00b | 5.01s | -710.00ms 🚀 |
| dist/16.9e97e787.js | 1.28kb | +0.00b | 1.89s | -394.00ms 🚀 |
| dist/16.acff901d.js | 1.27kb | +0.00b | 5.35s | +482.00ms ⚠️ |
| dist/component.233c7a65.js | 1.26kb | +0.00b | 3.42s | -473.00ms 🚀 |
| dist/16.015dfb9d.js | 1.26kb | +0.00b | 4.25s | -332.00ms 🚀 |
| dist/16.828b958e.js | 1.26kb | +0.00b | 4.59s | -497.00ms 🚀 |
| dist/16.b9b1643b.js | 1.26kb | +0.00b | 4.90s | -396.00ms 🚀 |
| dist/16.ec84a250.js | 1.26kb | +0.00b | 3.94s | -496.00ms 🚀 |
| dist/divider.ed7ccd2e.js | 1.26kb | +0.00b | 4.89s | -287.00ms 🚀 |
| dist/media-card-analytics-error-boundary.ff0177b3.js | 1.25kb | +0.00b | 11.27s | -824.00ms 🚀 |
| dist/16.458399e0.js | 1.23kb | +0.00b | 3.42s | -473.00ms 🚀 |
| dist/decision.c48f9f1e.js | 1.22kb | +0.00b | 4.89s | -284.00ms 🚀 |
| dist/panel-warning.b5f3360d.js | 1.22kb | +0.00b | 5.12s | -765.00ms 🚀 |
| dist/list.2dfe3f18.js | 1.19kb | +0.00b | 6.42s | +1.10s ⚠️ |
| dist/heading1.cc82b077.js | 1.19kb | +0.00b | 5.21s | -261.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.49b32f4d.js | 1023.00b | +0.00b | 11.83s | -3.11s 🚀 |
| dist/media-card-analytics-error-boundary.9a500f44.js | 1019.00b | +0.00b | 10.92s | +1.53s ⚠️ |
| dist/media-card-analytics-error-boundary.b884d580.js | 1019.00b | +0.00b | 12.38s | -1.50s 🚀 |
| dist/media-viewer-analytics-error-boundary.83da5db2.js | 1015.00b | +0.00b | 11.49s | +1.09s ⚠️ |
| dist/index.html | 119.00b | +0.00b | 2.37s | -576.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.e48a72e5.js | 2.46mb | +0.00b | 585.00ms | +59.00ms ⚠️ |
| dist/pdfRenderer.f5a3f4a7.js | 1.11mb | +0.00b | 616.00ms | +46.00ms ⚠️ |
| dist/editorView.ea13a2b7.js | 588.08kb | +0.00b | 640.00ms | +54.00ms ⚠️ |
| dist/popup.119ff481.js | 171.12kb | +0.00b | 641.00ms | +61.00ms ⚠️ |
| dist/Toolbar.0c83bb49.js | 100.68kb | +0.00b | 640.00ms | +54.00ms ⚠️ |
| dist/media-viewer.ebcb5e1e.js | 73.43kb | +0.00b | 608.00ms | +38.00ms ⚠️ |
| dist/card.df70c9b7.js | 54.67kb | +0.00b | 593.00ms | +31.00ms ⚠️ |
| dist/card.1d931b4c.js | 52.43kb | +0.00b | 624.00ms | +44.00ms ⚠️ |
| dist/Modal.558fb5db.js | 39.59kb | +0.00b | 584.00ms | +58.00ms ⚠️ |
| dist/component.ad27194c.js | 31.05kb | +0.00b | 584.00ms | +58.00ms ⚠️ |
| dist/smartMediaEditor.21f91ed8.js | 16.96kb | +0.00b | 640.00ms | +57.00ms ⚠️ |
| dist/dropzone.3ae9a030.js | 16.15kb | +0.00b | 616.00ms | +36.00ms ⚠️ |
| dist/card.3a4028bd.js | 5.74kb | +0.00b | 608.00ms | +38.00ms ⚠️ |
| dist/media-viewer.f15e56f4.js | 4.01kb | +0.00b | 641.00ms | +61.00ms ⚠️ |
| dist/dropzone.d1e8004d.js | 3.46kb | +0.00b | 641.00ms | +58.00ms ⚠️ |
| dist/index.ece3a13e.css | 3.46kb | +0.00b | 643.00ms | +54.00ms ⚠️ |
| dist/clipboard.6a6e5fbd.js | 3.00kb | +0.00b | 616.00ms | +36.00ms ⚠️ |
| dist/card.2888338b.js | 2.15kb | +0.00b | 641.00ms | +61.00ms ⚠️ |
| dist/feedback.facaac2e.js | 1.87kb | +0.00b | 582.00ms | +31.00ms ⚠️ |
| dist/browser.8883c452.js | 1.85kb | +0.00b | 616.00ms | +36.00ms ⚠️ |
| dist/workerHasher.d607e0b8.js | 1.77kb | +0.00b | 625.00ms | +42.00ms ⚠️ |
| dist/heading6.5a6082c8.js | 1.53kb | +0.00b | 581.00ms | +30.00ms ⚠️ |
| dist/heading5.88bc6249.js | 1.41kb | +0.00b | 581.00ms | +30.00ms ⚠️ |
| dist/expand.4a483eb0.js | 1.39kb | +0.00b | 582.00ms | +31.00ms ⚠️ |
| dist/heading4.a2524c70.js | 1.29kb | +0.00b | 580.00ms | +29.00ms |
| dist/component.233c7a65.js | 1.26kb | +0.00b | 585.00ms | +59.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.ff0177b3.js | 1.25kb | +0.00b | 608.00ms | +38.00ms ⚠️ |
| dist/16.458399e0.js | 1.23kb | +0.00b | 585.00ms | +59.00ms ⚠️ |
| dist/media-picker-analytics-error-boundary.49b32f4d.js | 1023.00b | +0.00b | 616.00ms | +36.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.9a500f44.js | 1019.00b | +0.00b | 608.00ms | +38.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.b884d580.js | 1019.00b | +0.00b | 624.00ms | +41.00ms ⚠️ |
| dist/media-viewer-analytics-error-boundary.83da5db2.js | 1015.00b | +0.00b | 641.00ms | +61.00ms ⚠️ |
| dist/simpleHasher.79137cb4.js | 767.00b | +0.00b | 590.00ms | +30.00ms ⚠️ |
| dist/simpleHasher.2c5206fd.js | 767.00b | +0.00b | 625.00ms | +42.00ms ⚠️ |
| dist/index.html | 119.00b | +0.00b | 501.00ms | +46.00ms ⚠️ |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 14.51s | -389.00ms |
| Cached | 574.00ms | +10.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.60kb | +0.00b | 90.00ms | -22.00ms 🚀 |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/296317007988523533)**

**Comment by joeyslater** [comment]
@mischnic @wbinnssmith - could I get some 👀 on this? I'd like some review before merging in v2 / resolving conflicts.

File: packages/core/core/src/ContentGraph.js
Comment (mischnic) [inline packages/core/core/src/ContentGraph.js:47]: Isn't `contentKey === node.id` always true? So I think the first argument could be removed

File: packages/core/core/src/ContentGraph.js
Comment (mischnic) [inline packages/core/core/src/ContentGraph.js]: I think this could be `return this._contentKeyToNodeId.has(contentKey)` instead?

File: packages/core/core/src/Parcel.js
Comment (mischnic) [inline packages/core/core/src/Parcel.js:322]: Seeing all of these changes makes me wonder if `getNode` could rather be `getNodeByContentKey` instead and then there's `getNodeByNodeId`.

But there are of course also benefits to directly using the node id and not constantly converting from the content id (e.g. in some bundle graph method)...

File: packages/core/core/src/AssetGraph.js
Comment (mischnic) [inline packages/core/core/src/AssetGraph.js]: nit: `replaceNodeIdsConnectedTo(..., nodes.map(node => this.addNode(node))); `

File: packages/core/core/src/AssetGraph.js
Comment (mischnic) [inline packages/core/core/src/AssetGraph.js]: nit: `replaceNodeIdsConnectedTo(..., resolved.map(node => this.addNode(nodeFromEntryFile(file))); `

File: packages/core/core/src/AssetGraph.js
Comment (mischnic) [inline packages/core/core/src/AssetGraph.js]: nit: use `.map`

File: packages/core/core/src/BundleGraph.js
Comment (mischnic) [inline packages/core/core/src/BundleGraph.js:131]: I would not remove that assertion for the root node to exist

File: packages/core/core/src/BundleGraph.js
Comment (mischnic) [inline packages/core/core/src/BundleGraph.js:528]: Wouldn't the call to `removeNode` have thrown previously anyway? (Because of the assertion in `removeNode`)
Why is the condition needed now?

File: packages/core/core/src/BundleGraph.js
Comment (mischnic) [inline packages/core/core/src/BundleGraph.js:722]: There are a few places where `getNodesConnectedTo` is now replaced with `getNodeIdsConnectedTo().map(id => getNode(id))`. Maybe keep `getNodesConnectedTo`?

File: packages/core/core/src/BundleGraph.js
Comment (mischnic) [inline packages/core/core/src/BundleGraph.js]: ...

File: packages/core/core/src/Graph.js
Comment (mischnic) [inline packages/core/core/src/Graph.js]: `AssetGraph` uses this method, so it's not really `_private`?

File: packages/core/core/src/requests/AssetGraphRequest.js
Comment (mischnic) [inline packages/core/core/src/requests/AssetGraphRequest.js]:
```suggestion
      > = assetNode.value.symbols;
```

File: packages/core/core/src/requests/AssetGraphRequest.js
Comment (mischnic) [inline packages/core/core/src/requests/AssetGraphRequest.js]:
```suggestion
        let connectedNodes = this.assetGraph.getNodeIdsConnectedTo(
```

File: packages/core/core/src/requests/AssetGraphRequest.js
Comment (mischnic) [inline packages/core/core/src/requests/AssetGraphRequest.js]: Can you pass in the `AssetNode` as before here and retrieve the node in `propagateSymbolsUp` (and not in the callback)?

File: packages/core/core/src/requests/AssetGraphRequest.js
Comment (mischnic) [inline packages/core/core/src/requests/AssetGraphRequest.js]: Can you pass in the `AssetNode` as before here and retrieve the node in `propagateSymbolsDown` (and not in the callback)?

File: packages/core/core/src/Graph.js
Comment (mischnic) [inline packages/core/core/src/Graph.js:95]: Do we want to have a `graph.setRootNodeId(id: NodeId)` method instead of doing `graph.rootId = ...`?

File: packages/core/core/src/RequestTracker.js
Comment (mischnic) [inline packages/core/core/src/RequestTracker.js]: There are many occurrences of this pattern in this file, maybe this could become `graph.ensureNode(node: TNode): NodeID`

File: packages/core/core/src/RequestTracker.js
Comment (mischnic) [inline packages/core/core/src/RequestTracker.js]: nit: put `'file_name:' + basename` into a variable (similar to how it was specified just once previously)

File: packages/core/core/src/RequestTracker.js
Comment (mischnic) [inline packages/core/core/src/RequestTracker.js]: Using the "private" variable...

File: packages/core/core/src/ContentGraph.js
Comment (joeyslater) [reply packages/core/core/src/ContentGraph.js:47]: I believe we want to eventually remove id from Node since the id on node is the node value's id. @wbinnssmith , that's what we talked through, right?

File: packages/core/core/src/BundleGraph.js
Comment (joeyslater) [reply packages/core/core/src/BundleGraph.js:528]: I believe this check was added because the issue was being thrown, so this prevents the case of being thrown if the node was revisited by the removal chain.

File: packages/core/core/src/RequestTracker.js
Comment (mischnic) [reply packages/core/core/src/RequestTracker.js]: Actually, should this rather be included in `addNode`? At the moment you could call `addNode(foo);` twice and the node will actually be inserted twice

Comment (joeyslater) [reply packages/core/core/src/RequestTracker.js]: I am not entirely sure of the logic of the request graph, but if I change this to addNode - the tests fail. It seems like when `addNode` overrides the current node value.

Comment (mischnic) [reply packages/core/core/src/RequestTracker.js]: I was wondering whether `addNode` should actually behave like `ensureNode`, because this behaviour sounds error prone:

> At the moment you could call addNode(foo); twice and the node will actually be inserted twice

**Review by devongovett — APPROVED** [review]
There's a lot here, but lgtm from my skim. I'll let @mischnic approve as well since he reviewed previously.

**Comment by mischnic** [comment]
The two changes to `AssetGraphRequest.js` I commented on would be great, the other comments are just thoughts I had (e.g. what should happen when `addNode` is called twice with the same object/same content id).

**Comment by joeyslater** [comment]
> The two changes to `AssetGraphRequest.js` I commented on would be great, the other comments are just thoughts I had (e.g. what should happen when `addNode` is called twice with the same object/same content id).

AssetGraphRequest changes have been made and updated the addNode in RequestGraph. Everything was passing. Merged in v2, so hopefully the pipeline will be green and ready to go

**Review by mischnic — APPROVED** [review]

**Pull request merged**
